pandas/core/arrays/datetimes.py

```diff
@@ -290,9 +290,15 @@ def _scalar_type(self) -> type[Timestamp]:
 
     @classmethod
     def _from_scalars(cls, scalars, *, dtype: DtypeObj) -> Self:
-        if lib.infer_dtype(scalars, skipna=True) not in ["datetime", "datetime64"]:
-            # TODO: require any NAs be valid-for-DTA
-            # TODO: if dtype is passed, check for tzawareness compat?
+        # TODO: require any NAs be valid-for-DTA
+        # TODO: if dtype is passed, check for tzawareness compat?
+        if not lib.is_datetime64_array(scalars):
```

Comment:
I believe we need this (or something like it) because e.g. `lib.is_datetime_with_singletz_array` does not check if there are datetimes:

```
print(lib.is_datetime_with_singletz_array(np.array([1, 2, 3])))
# True
```

However, I do not currently understand the behavior differences between this and the previous `lib.infer_dtype` and need to look into this more. Any guidance here is much appreciated.

Reply:
Looks like we get that behavior from is_datetime_with_singletz_array bc it implicitly assumes it is called from maybe_convert_objects, where it will only get called if the first non-NaT it sees is a datetime with a tz.

I'm not a fan of infer_dtype bc the string returns are clunky (and hard to reason about, e.g. when do you expect "datetime" vs "datetime64") and fragile when we add new return values (e.g. "integer-na").  Also in cases when you only care about datetimes, it will not short-circuit when it sees a non-datetime.

Reply:
Thanks - makes perfect sense to move away from `infer_dtype`. For the old behavior, I'm seeing the use of `infer_dtype` allow the following to not raise on a non-empty input:

 - If the `scalars` has the attribute `type`, `kind`, `name`, or `base` set to one of `"datetime64[ns]"`, `"M"`, `datetime`
 - `is_datetime64_array` returns True
 - `is_datetime_array` returns True

As far as I can tell, the input `scalars` here must be an ndarray, Index, or ExtensionArray, but I don't feel confident about this from the code path in `Index.map`.

With this, I'm wondering if the following is a good equivalent:

```
if (
    not is_datetime64_any_dtype(scalars)
    and not lib.is_datetime64_array(scalars)
    and not lib.is_datetime_array(scalars)
):
    raise ValueError
```